Code/Framework/AzCore/AzCore/RTTI/BehaviorContext.h

```diff
@@ -206,25 +216,48 @@ namespace AZ
     public:
         AZ_CLASS_ALLOCATOR(BehaviorDefaultValue, AZ::SystemAllocator, 0);
 
+        const BehaviorValueParameter& GetValue() const
+        {
+            return m_value;
+        }
+
+        BehaviorValueParameter m_value;
+    };
+
+    template<typename Value>
+    class BehaviorDefaultValueTyped
+        : public BehaviorDefaultValue
+    {
+    public:
         /**
         * Create a default value for a specific method parameter. The Default values is stored by value
         * in a temp storage, so currently there is limit the \ref BehaviorValueParameter temp storage, we
         * can easily change that if it became a problem.
         */
-        template<typename Value>
-        BehaviorDefaultValue(Value&& value)
+        BehaviorDefaultValueTyped(Value&& value)
         {
-            m_value.StoreInTempData(AZStd::forward<Value>(value));
+            m_value.StoreInTempDataEvenIfNonTrivial(AZStd::forward<Value>(value));
         }
 
-        const BehaviorValueParameter& GetValue() const
+        ~BehaviorDefaultValueTyped() override
         {
-            return m_value;
+            if constexpr (!AZStd::is_trivially_destructible_v<AZStd::decay_t<Value>>)
+            {
+                // BehaviorValueParameter does not destroy instances of types set
+                // with StoreInTempData(). It does not take ownership of that
+                // instance, because it is designed to reference existing values on
+                // the stack. For a parameter's default value, however, that
+                // instance must be owned by something, so that ownership is the
+                // responsibility of the BehaviorDefaultValue type. If the type of
+                // the default value has a non-trivial destructor, this class will
+                // invoke that destructor when it itself is destroyed.
+                AZStd::destroy_at(reinterpret_cast<AZStd::decay_t<Value>*>(m_value.m_value));
+            }
```

Comment:
This also check `m_value.m_value` is not nullptr before invoking the destructor on it.
Someone could create a default value parameter with a value that is nullptr.
Such as `behaviorContext->MakeDefaultValue(static_cast<int*>(nullptr))`